README.md
# Web-based Spoon (aka webSpoon)

webSpoon is a web-based graphical designer for Pentaho Data Integration with the same look & feel as Spoon.
Kettle transformation/job files can be designed and executed in your favorite web browser.
This is one of the community activities and not supported by Pentaho.

## Use cases

- Data security
- Remote use
- Ease of management
- Cloud

Please see [public talks](https://www.slideshare.net/HiromuHota/presentations) for more details.

# How to use

Please refer to the [wiki](https://github.com/HiromuHota/pentaho-kettle/wiki) and [issues](https://github.com/HiromuHota/pentaho-kettle/issues).

# How to deploy

There are two ways: with Docker and without Docker.
Docker is recommended because it is simple hence error-free.
In either way, webSpoon will be deployed at `http://address:8080/spoon/spoon`.

## With Docker (recommended)

```
$ docker run -d -p 8080:8080 hiromuhota/webspoon:latest-full
```

## Without Docker

Please refer to the [wiki](https://github.com/HiromuHota/pentaho-kettle/wiki/System-Requirements) for system requirements.

1. Unzip `pdi-ce-8.1.0.0-365.zip`, then copy `system` and `plugins` folders to `$CATALINA_HOME`.
2. Run install.sh at `$CATALINA_HOME`.
3. (Re)start the Tomcat.

The actual commands look like below:

```
$ export CATALINA_HOME=/home/vagrant/apache-tomcat-8.5.23
$ cd ~/
$ unzip ~/Downloads/pdi-ce-8.1.0.0-365.zip
$ cd $CATALINA_HOME
$ cp -r ~/data-integration/system ./
$ cp -r ~/data-integration/plugins ./
$ wget https://raw.githubusercontent.com/HiromuHota/webspoon-docker/0.8.1.17/install.sh
$ chmod +x install.sh
$ export version=0.8.1.17
$ export dist=8.1.0.0-365
$ ./install.sh
$ ./bin/startup.sh
```

Please use the right version of install.sh corresponding to your `version`.

# How to config (optional)

## Carte

If `$CATALINA_HOME/system/kettle/slave-server-config.xml` exists, the embedded Carte servlet can be configured accordingly.
See [here](https://wiki.pentaho.com/display/EAI/Carte+Configuration) for the config format.
An example config xml looks like this:

```
<slave_config>
  <max_log_lines>10000</max_log_lines>
  <max_log_timeout_minutes>2880</max_log_timeout_minutes>
  <object_timeout_minutes>240</object_timeout_minutes>
  <repository>
    <name>test</name>
    <username>username</username>
    <password>password</password>
  </repository>
</slave_config>
```

Note that only repositories defined in `$HOME/.kettle/repositories.xml` can be used.
Also note that it is NOT recommended to define `repository` element in the XML when the following user authentication is enabled.
Issues such as [#91](https://github.com/HiromuHota/pentaho-kettle/issues/91) can happen.

## User authentication

Edit `WEB-INF/web.xml` to uncomment/enable user authentication.

```
  <!-- Uncomment the followings to enable login page for webSpoon
  <context-param>
    <param-name>contextConfigLocation</param-name>
    <param-value>
      /WEB-INF/spring/*.xml
    </param-value>
  </context-param>

  <filter>
    <filter-name>springSecurityFilterChain</filter-name>
    <filter-class>org.springframework.web.filter.DelegatingFilterProxy</filter-class>
  </filter>
  <filter-mapping>
    <filter-name>springSecurityFilterChain</filter-name>
    <url-pattern>/*</url-pattern>
  </filter-mapping>

  <listener>
    <listener-class>org.springframework.web.context.ContextLoaderListener</listener-class>
  </listener>
  -->
```

Edit `WEB-INF/spring/security.xml` to configure authentication providers.
The examle below uses two auth providers: `ldap-authentication-provider` and `authentication-provider`, where the LDAP authentication is provided by the Spring Security's embedded LDAP server with a configuration of `WEB-INF/classes/example.ldif`.
See [here](https://docs.spring.io/spring-security/site/docs/4.1.x/reference/html/ns-config.html) for Spring Security in general and [here](http://docs.spring.io/spring-security/site/docs/4.1.x/reference/html/ldap.html) for LDAP.

```
  <!--<ldap-server url="ldap://localhost:389/dc=example,dc=org"/>-->
  <ldap-server ldif="classpath:example.ldif" root="dc=example,dc=org"/>

  <authentication-manager>
    <ldap-authentication-provider user-dn-pattern="uid={0},ou=people"
      group-search-base="ou=groups" />
    <authentication-provider>
      <user-service>
        <user name="user" password="password" authorities="ROLE_USER" />
      </user-service>
    </authentication-provider>
  </authentication-manager>
```

## Third-party plugins and JDBC drivers

Place third-party plugins in `$CATALINA_HOME/plugins` and JDBC drivers in `$CATALINA_HOME/lib` as below:

```
$CATALINA_HOME
├── system
├── plugins
│   ├── YourPlugin
│   │   └── YourPlugin.jar
│   ├── ...
├── lib
│   ├── YourJDBC.jar
│   ├── ...
├── webapps
│   ├── spoon.war
│   ├── ...
```

## Disable UI component

Spoon uses the XUL (XML User Interface Language) to define some parts of its user interface (see [here](https://wiki.pentaho.com/display/ServerDoc2x/The+Pentaho+XUL+Framework+Developer%27s+Guide) for details).
The file menu, for example, is defined in `webapps/spoon/WEB-INF/classes/ui/menubar.xul` as in the following snippet:

```
<menu id="file" label="${Spoon.Menu.File}" accesskey="alt-f">
  <menuitem id="file-open" label="${Spoon.Menu.File.Open}" />
  <menuitem id="file-save-as" label="${Spoon.Menu.File.SaveAs}" />
</menu>
```

To restrict user's capability, one may want to disable some of UI components.
To to so, add `disable="true"` to the component to be disabled like below.

```
<menu id="file" label="${Spoon.Menu.File}" accesskey="alt-f">
  <menuitem id="file-open" label="${Spoon.Menu.File.Open}" />
  <menuitem id="file-save-as" label="${Spoon.Menu.File.SaveAs}" disabled="true" />
</menu>
```

No `disabled` attribute has the same effect as `disabled="false"`.

# How to develop

Spoon relies on SWT for UI widgets, which is great for being OS agnostic, but it only runs as a desktop app.
RAP/RWT provides web UIs with SWT API, so replacing SWT with RAP/RWT allows Spoon to run as a web app with a little code change.
Having said that, some APIs are not implemented; hence, a little more code change is required than it sounds.

## Design philosophy

1. Minimize the difference from the original Spoon.
2. Optimize webSpoon as a web application.

## Branches and Versioning

I started this project in the webspoon branch, branched off from the branch 6.1 of between 6.1.0.5-R and 6.1.0.6-R.
Soon I realized that I should have branched off from one of released versions.
So I decided to make two branches: webspoon-6.1 and webspoon-7.0, each of which was rebased onto 6.1.0.1-R and 7.0.0.0-R, respectively.

webSpoon uses 4 digits versioning with the following rules:

- The 1st digit is always 0 (never be released as a separate software).
- The 2nd and 3rd digits represent the base Kettle version, e.g., 6.1, 7.0.
- The last digit represents the patch version.

As a result, the next (pre-)release version will be 0.6.1.4, meaning it is based on the Kettle version 6.1 with the 4th patch.
There could be a version of 0.7.0.4, which is based on the Kettle version 7.0 with (basically) the same patch.

## Build and locally publish dependent libraries

Please build and locally-publish the following dependent libraries.

- pentaho-xul-swt
- org.eclipse.rap.rwt
- org.eclipse.rap.jface
- org.eclipse.rap.fileupload
- org.eclipse.rap.filedialog
- org.eclipse.rap.rwt.testfixture
- pentaho-vfs-browser

### pentaho-xul-swt

```
$ git clone -b webspoon-8.1 https://github.com/HiromuHota/pentaho-commons-xul.git
$ cd pentaho-commons-xul
$ mvn clean install -pl swt
```

### rap

```
$ git clone -b webspoon-3.1-maintenance https://github.com/HiromuHota/rap.git
$ cd rap
$ mvn clean install -N
$ mvn clean install -pl bundles/org.eclipse.rap.rwt -am
$ mvn clean install -pl bundles/org.eclipse.rap.jface -am
$ mvn clean install -pl bundles/org.eclipse.rap.fileupload -am
$ mvn clean install -pl bundles/org.eclipse.rap.filedialog -am
$ mvn clean install -pl tests/org.eclipse.rap.rwt.testfixture -am
```

### pentaho-vfs-browser

```
$ git clone -b webspoon-8.1 https://github.com/HiromuHota/apache-vfs-browser.git
$ cd apache-vfs-browser
$ mvn clean install
```

Remember to build pentaho-vfs-browser after rap due to its dependency on rap.

## Build in the command line

**Make sure patched dependent libraries have been published locally**

Compile `kettle-core`, `kettle-engine`, `kettle-ui-swt`, and `webspoon-security`; and install the packages into the local repository:

```bash
$ git clone -b webspoon-8.1 https://github.com/HiromuHota/pentaho-kettle.git
$ cd pentaho-kettle
$ mvn clean install -pl core,engine,security,ui
```

Optionally you can specify -Dmaven.test.skip=true to skip the tests

Build a war file (`spoon.war`):

```bash
$ mvn clean install -pl assemblies/static
$ mvn clean install -pl assemblies/lib
$ mvn clean package -pl assemblies/client
```

## UI testing using Selenium

Currently, only Google Chrome browser has been tested for when running UI test cases.
The tests run in headless mode unless a parameter `-Dheadless.unittest=false` is passed.
To run tests in headless mode, the version of Chrome should be higher than 59.

The default url is `http://localhost:8080/spoon`.
Pass a parameter like below if webSpoon is deployed to a different url.

The following command runs all the unit test cases including UI in non-headless mode.

```
$ mvn clean test -pl integration -Dtest.baseurl=http://localhost:8080/spoon/spoon -Dheadless.unittest=false
```

# Notices

- Pentaho is a registered trademark of Hitachi Vantara Corporation.
- Google Chrome browser is a trademark of Google Inc.
- Other company and product names mentioned in this document may be the trademarks of their respective owners.


CHANGELOG.md
# Change Log

## 0.8.1.17 - 2018-11-26
### Changed
- Update RAP/RWT to 3.6.0
- Update Selenium version to 3.14.0

### Fixed
- Fix popup menus in Filter Rows (fix #112)
- Log the current directory path thru JUL (fix #116)
- Avoid static modifier for session unique instance (fix #117)
- Use internalization key instead of hardcode (thanks @Valeran86)
- Open a help page in a separate tab (fix #120)
- Cancel the traverse event invoked by TAB (+SHIFT) key (fix #123)
- Explicitly set a new menu to a control (fix #122)
- Export To XML with its original name instead of a random string (fix #124)
- Adjust the dialog height (fix #125)
- saveXMLFileToVfs uses meta.getName() instead of "Untitled" (fix #128)

## 0.8.1.16 - 2018-06-29
### Changed
- Rebased to 8.1.0.0-365

## 0.8.0.16 - 2018-05-17
### Added
- Include WEB-INF/classes/log4j.xml to configure log4j logging framework
- Support the system clipboard
- Restored "Snapshot Canvas"

### Changed
- Let logger "org.pentaho.di" write out to logs/pdi.log
- Log4j writes to spoon.log unless otherwise configured like "org.pentaho.di"
- Change the logging framework for Carte config to Log4j
- Use spring-tx which supersedes spring-dao as of 2.5

### Removed
- Remove apacheds-all to resolve version conflicts of slf4j
- Remove the example ldif file (fix #102)

### Fixed
- [#103] Unintentional triggering of the selection mode

### Security
- vulnerable component spring-security: upgrade from 4.1.3 to 4.1.5 (https://pivotal.io/security/cve-2018-1199)
- Update vulnerable component: Spring Ldap Core (https://pivotal.io/security/cve-2017-8028)
- Update vulnerable component: Spring Framework (https://pivotal.io/security/cve-2018-1275)

## 0.8.0.15 - 2018-03-19
### Added
- Canvas drawing instructions at client-side to improve interactivity
- Support attribute "disabled" to turn menuitems on/off

### Changed
- No drawGrid instructions from the server-side to improve responsiveness
- Disable the capability manager and kettle.properties editor only when they are disabled="true" in menubar.xul

### Fixed
- [#99] Clear the entire canvas before drawing

## 0.8.0.14 - 2018-02-07
### Added
- Add a custom security manager (disabled by default)
- Add jars required for LDAP authentication and example .ldif file

### Changed
- Update RAP/RWT to 3.3.0
- Exclude jetty jars
- Add pentaho-platform-scheduler
- Add jersey-spring
- Add required spring jars for user auth
- [#90] Configure Carte if slave-server-config.xml exists
- Add the patched pentaho-vfs-browser as a dependency
- Secure Carte endpoints by BASIC-Auth
- Move user-service element under authentication-manager/authentication-provider

### Fixed
- [#45] Fixed the popupmenu location (by the patched pentaho-vfs-browser)
- [#85] Fixed "No suitable driver found for Hive2/Impala" error
- [#86] Fixed "NoClassDefFoundError: org/eclipse/jface/window/DefaultToolTip" error
- [#88] Include commons-fileupload (required for "Import from an XML file")
- [#92] Fixed "UI session stops working after 500 (Internal Server Error)"
- [#94] Adding `save as` behaviour for `FileDialog.open()` (Thanks to @Valeran86)
- [HiromuHota/rap#1] Set correct message on finish button in Wizard (Thanks to @Valeran86)

## 0.8.0.13 - 2017-11-17
### Changed
- kettle-engine needs patching
- Move WebSpoonTest.java to integration

## 0.7.1.13 - 2017-10-20

### Changed
- Make repositories-plugin multi-user enabled

### Fixed
- [#58] Cursor position of a script editor is not accurate
- [#75] File > Open... does not open a Kettle file
- [#78] Checkboxes not drawn for the database connection pooling parameters
- [#80] Location is not reflected in the Repository Manager dialog
- Use logError instead of ErrorDialog b/c shell is not yet available in the Spoon constructor

## 0.7.1.12 - 2017-08-16

### Added
- Add some more Selenium UI tests to detect regressions caused by version upgrade

### Changed
- Eable open/save menus even when not connected to a repository
- Firewall / port forward for the OSGI Service (e.g., 9051) is no longer required

### Fixed
- Restored the Repository dialog that was missing when deployed to a url other than localhost
- Upgrade xalan from 2.6.0 to 2.7.2 to resolve the issue that Karaf not starting up when webSpoon deployed in a Docker container

## 0.7.1.11 - 2017-07-14

### Changed
- More multi-user friendly by assigning `.kettle` and `.pentaho/metastore` to each user
- VfsFileChooserDialog for FileDialog
- Move OSGi HTTP service to /osgi (to align with Pentaho Server)
- Move ui/* from WEB-INF/classes/ui/ to the inside of kettle-ui-swt-XXX.jar
- Create webSpoon\_OSS\_Licenses.html for webSpoon-specific OSS
- ShowExitWarning=Y by default and respect the property
- Disable the capability manager and kettle.properties editor when used by multi-user

### Fixed
- Metrics tabs in Test Data Service dialog
- PDI-14492: Copy Table Wizard_UI Issue

## 0.7.1.10 - 2017-06-14

### Changed
- VfsFileChooserDialog for DirectoryDialog
- Use headless Chrome for UI testing
- Disable keyboard shortcut F4 because it is invoked when Enter key is pressed for unknown reason
- Disable keyboard shortcuts under "Edit" menu while a dialog (e.g., step dialog) is focused
- Do not share logs across sessions and do not show the general logs

### Removed
- Remove the Timer for setControlStates() and execute it when trans/job Finished
- Stop using DelayTimer because it does not effectively work in webSpoon

### Fixed
- Make RunConfigurationPopupMenu multi-session enabled
- Fix the "SWTError: BrowserFunction is disposed" when pressing Connect Now
- Refresh the log display periodically
- Restore the ability to connect to a repository by query parameters
- Restore the connectivity to Pentaho Repository and Database Repository
- Catch any exception when executing trans.startThreads() or job.start()
- Move steps/job entries only if a cursor moves

## 0.7.1.9 - 2017-05-23

### Changed
- Rebased to 7.1.0.0-12

## 0.7.0.9 - 2017-05-05

### Added

ui/src/main/java/org/pentaho/di/ui/core/ConstUI.java
/*! ******************************************************************************
 *
 * Pentaho Data Integration
 *
 * Copyright (C) 2002-2018 by Hitachi Vantara : http://www.pentaho.com
 *
 *******************************************************************************
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************************/

package org.pentaho.di.ui.core;

import org.eclipse.jface.action.MenuManager;
import org.eclipse.swt.widgets.Control;
import org.eclipse.swt.widgets.Menu;
import org.eclipse.swt.widgets.TreeItem;
import org.pentaho.di.core.Const;
import org.pentaho.di.core.util.Utils;
import org.pentaho.ui.xul.containers.XulMenupopup;

/**
 * This class is used to define a number of default values for various settings throughout Kettle. It also contains a
 * number of static final methods to make your life easier.
 *
 * @author Matt
 * @since 07-05-2003
 *
 */
public class ConstUI {
  /**
   * Default icon size
   */
  public static final int ICON_SIZE = 32;

  public static final int LARGE_ICON_SIZE = 48;

  public static final int SMALL_ICON_SIZE = 16;

  public static final int MEDIUM_ICON_SIZE = 24;

  /**
   * Default checkbox width
   */
  public static final int CHECKBOX_WIDTH = 20;

  /**
   * Default line width for arrows & around icons
   */
  public static final int LINE_WIDTH = 1;

  /**
   * Default grid size to which the graphical views snap.
   */
  public static final int GRID_SIZE = 16;

  /**
   * The minimal size of a note on a graphical view (width & height)
   */
  public static final int NOTE_MIN_SIZE = 20;

  /**
   * Offset between pointer and tooltip position.
   */
  public static final int TOOLTIP_OFFSET = 5;

  /**
   * The default red-component of the background color
   */
  public static final int COLOR_BACKGROUND_RED = 255;

  /**
   * The default green-component of the background color
   */
  public static final int COLOR_BACKGROUND_GREEN = 255;

  /**
   * The default blue-component of the background color
   */
  public static final int COLOR_BACKGROUND_BLUE = 255;

  /**
   * The default red-component of the graph background color
   */
  public static final int COLOR_GRAPH_RED = 255;

  /**
   * The default green-component of the graph background color
   */
  public static final int COLOR_GRAPH_GREEN = 255;

  /**
   * The default blue-component of the graph background color
   */
  public static final int COLOR_GRAPH_BLUE = 255;

  /**
   * The default red-component of the tab selected color
   */
  public static final int COLOR_TAB_RED = 240;

  /**
   * The default green-component of the tab selected color
   */
  public static final int COLOR_TAB_GREEN = 240;

  /**
   * The default blue-component of the tab selected color
   */
  public static final int COLOR_TAB_BLUE = 240;

  /**
   * the default canvas refresh interval for running transformations
   */
  public static final int INTERVAL_MS_TRANS_CANVAS_REFRESH = 1000;

  /**
   * margin between points of controls
   */
  public static final int SMALL_MARGIN = 5;

  /**
   * margin between points of controls
   */
  public static final int MEDUIM_MARGIN = 10;

  /**
   * Determine the level of where the TreeItem is position in a tree.
   *
   * @param ti
   *          The TreeItem
   * @return The level of the item in the tree
   */
  public static final int getTreeLevel( TreeItem ti ) {
    int level = 0;
    TreeItem parent = ti.getParentItem();
    while ( parent != null ) {
      level++;
      parent = parent.getParentItem();
    }

    return level;
  }

  /**
   * Get an array of strings containing the path from the given TreeItem to the parent.
   *
   * @param ti
   *          The TreeItem to look at
   * @return An array of string describing the path to the TreeItem.
   */
  public static final String[] getTreeStrings( TreeItem ti ) {
    int nrlevels = getTreeLevel( ti ) + 1;
    String[] retval = new String[nrlevels];
    int level = 0;

    retval[nrlevels - 1] = ti.getText();
    TreeItem parent = ti.getParentItem();
    while ( parent != null ) {
      level++;
      retval[nrlevels - level - 1] = parent.getText();
      parent = parent.getParentItem();
    }

    return retval;
  }

  /**
   * Return the tree path seperated by Const.FILE_SEPARATOR, starting from a certain depth in the tree.
   *
   * @param ti
   *          The TreeItem to get the path for
   * @param from
   *          The depth to start at, use 0 to get the complete tree.
   * @return The tree path.
   */
  public static final String getTreePath( TreeItem ti, int from ) {
    String[] path = getTreeStrings( ti );

    if ( path == null ) {
      return null;
    }

    String retval = "";

    for ( int i = from; i < path.length; i++ ) {
      if ( !path[i].equalsIgnoreCase( Const.FILE_SEPARATOR ) ) {
        retval += Const.FILE_SEPARATOR + path[i];
      }
    }

    return retval;
  }

  /**
   * Flips the TreeItem from expanded to not expanded or vice-versa.
   *
   * @param ti
   *          The TreeItem to flip.
   */
  public static final void flipExpanded( TreeItem ti ) {
    ti.setExpanded( !ti.getExpanded() );
  }

  public static final TreeItem findTreeItem( TreeItem parent, String name ) {
    return findTreeItem( parent, null, name );
  }

  /**
   * Finds a TreeItem with a certain label (name) in a (part of a) tree.
   *
   * @param parent
   *          The TreeItem where we start looking.
   * @param parentName
   *          The name of the parent to match as well (null=not used)
   * @param name
   *          The name or item label to look for.
   * @return The TreeItem if the label was found, null if nothing was found.
   */
  public static final TreeItem findTreeItem( TreeItem parent, String parentName, String name ) {
    return findTreeItem( null, parent, parentName, name );
  }

  private static final TreeItem findTreeItem( TreeItem grandParent, TreeItem parent, String parentName, String name ) {
    if ( Utils.isEmpty( parentName ) ) {
      if ( parent.getText().equalsIgnoreCase( name ) ) {
        return parent;
      }
    } else {
      if ( grandParent != null && grandParent.getText().equalsIgnoreCase( "OTHER" ) ) {
        System.out.println( "Other" );
      }
      if ( grandParent != null
        && grandParent.getText().equalsIgnoreCase( parentName ) && parent.getText().equalsIgnoreCase( name ) ) {
        return parent;
      }
    }

    TreeItem[] ti = parent.getItems();
    for ( int i = 0; i < ti.length; i++ ) {
      TreeItem child = findTreeItem( parent, ti[i], parentName, name );
      if ( child != null ) {
        return child;
      }
    }
    return null;
  }

  public static void displayMenu( XulMenupopup xulMenuPopup, Control control ) {

    MenuManager menuMgr = (MenuManager) xulMenuPopup.getManagedObject();
    Menu menu = menuMgr.createContextMenu( control );
    menuMgr.updateAll( true );
    displayMenu( menu, control );
  }

  public static void displayMenu( Menu menu, Control control ) {
    Menu oldMenu = control.getMenu();
    if ( oldMenu != null && oldMenu != menu ) {
      oldMenu.setVisible( false );
    }

    // XXX: Stubbing out this line prevents context dialogs from appearing twice
    // on OS X. Tested on Windows to be sure there is no adverse effect.
    // Unfortunately, I do *not* understand why this works. I ran it by
    // mcasters and he didn't know for sure either.
    // HiromuHota: uncomment this line to fix HiromuHota/pentaho-kettle#122
    control.setMenu( menu );
    menu.setLocation( control.getDisplay().getCursorLocation() );
    menu.setVisible( true );
  }

}
